### PR TITLE
Allow NU_LIBS_DIR and friends to be const

### PR DESCRIPTION
**Pull request by StevenDoesStuffs**
# Description

Allow NU_LIBS_DIR and friends to be const they can be updated within the same parse pass. This will allow us to remove having multiple config files eventually. 

Small implementation detail: I've changed `call.parser_info` to a hashmap with string keys, so the information can have names rather than indices, and we don't have to worry too much about the order in which we put things into it.

Closes https://github.com/nushell/nushell/issues/8422

# User-Facing Changes

In a single file, users can now do stuff like
```
const NU_LIBS_DIR = ['/some/path/here']
source script.nu
```
and the source statement will use the value of NU_LIBS_DIR declared the line before. 

Currently, if there is no `NU_LIBS_DIR` const, then we fallback to using the value of the `NU_LIBS_DIR` env-var, so there are no breaking changes (unless someone named a const NU_LIBS_DIR for some reason). 

2023-03-04-014103_hyprshot

# Tests + Formatting

~~TODO: write tests~~ Done

# After Submitting

~~TODO: update docs~~ Will do when we update default_env.nu/merge default_env.nu into default_config.nu.

### CODE REVIEW

File: crates/nu-parser/src/parse_keywords.rs [hunk 1]
```diff
@@ -4006,52 +4020,57 @@ pub fn parse_register(
...
+            let dir_abs = canonicalize_with(dir, actual_cwd).ok()?;
+            canonicalize_with(filename, dir_abs).ok()
+        })
+        .find(Option::is_some)
+        .flatten()
+}
+
+pub fn find_dirs_var(working_set: &StateWorkingSet, var_name: &str) -> Expr {
+    working_set
+        .find_variable(format!("${}", var_name).as_bytes())
```
Comment (kubouch) [inline crates/nu-parser/src/parse_keywords.rs:4072]: I don't think you need to prepend the `$`. Does just `find_variable(var_name.as_bytes())` work?

Comment (StevenDoesStuffs) [reply crates/nu-parser/src/parse_keywords.rs:4072]: Unfortunately not. I had it without prepending $ at first and it didn't work until I added it.

**Comment by kubouch** [comment]
OK, cool, seems mostly good but there might be some unnecessary complexity: We don't need to keep `$env.NU_LIB_DIRS` around. It seems to complicate the logic of `find_dirs_var()` which IMO should just look for a constant and return an Option. Backwards compatibility could be done by an ad-hoc check that can be easily removed the next release, or we can just print a warning in `Stack.add_env_var()` if you try to set NU_LIB/PLUGIN_DIRS env vars.

Also, the default env https://github.com/nushell/nushell/blob/main/crates/nu-utils/src/sample_config/default_env.nu should reflect change NU_LIB/PLUGIN_DIRS to a constant.

**Comment by kubouch** [comment]
After merging this PR, we can merge the default_env.nu and default_config.nu into one file (default_config.nu). Also, after merging your other PR, we can remove the login.nu. But this would better be left for another PR, this one is big enough.

File: crates/nu-parser/src/parse_keywords.rs (hunk 1 above)
Comment (fdncred) [reply crates/nu-parser/src/parse_keywords.rs:4072]: I'm guessing that this is because if you do `$nu.scope.var` all the variables start with `$`. Another way to fix this is to take the `$` off when they're being created and stored. I never really understood why we were artificially prepending a `$` in the first place, but there could be a valid reason.

You can see where we add it here
https://github.com/nushell/nushell/blob/26c36e932e62728b7930643c2f1b3ef0e906ae36/crates/nu-protocol/src/engine/engine_state.rs#L1807-L1810

This would probably have to change too
https://github.com/nushell/nushell/blob/26c36e932e62728b7930643c2f1b3ef0e906ae36/crates/nu-parser/src/parser.rs#L117-L123

Comment (kubouch) [reply crates/nu-parser/src/parse_keywords.rs:4072]: I see. Yeah, seems unnecessary to keep adding the `$` and can probably be removed, but also it's quite off-topic for this PR so it's not worth spending too much time on. It could be fixed in another PR.

Comment (fdncred) [reply crates/nu-parser/src/parse_keywords.rs:4072]: Agreed. Just wanted to bring it up since I found it recently doing other `$` stuff.

**Comment by StevenDoesStuffs** [comment]
> Backwards compatibility could be done by an ad-hoc check that can be easily removed the next release

I'm not quite sure I understand what this means. Do you mean that we can move the compat logic into find_env_dirs?

**Comment by kubouch** [comment]
What I meant is that the `find_dirs_var()` returns `Expr::String` if it's not a constant which is then used for env var lookup which makes it a bit hard to follow. This can be simplified. I'm not sure what is the `find_env_dirs` you refer to, but the old `find_in_dirs()` could be named as `find_in_dirs_old()` which would look for the `$env.NU_LIB_DIRS`. You would check both `find_in_dirs()` and `find_in_dirs_old()`. The latter  and would be deleted after the release. We can also just remove support for `$env.NU_LIB_DIRS` entirely, and just add a warning when setting the env var as I mentioned above.

**Comment by StevenDoesStuffs** [comment]
Ah sorry I meant find_in_dirs. Anyhow I'll make the changes soon.

**Comment by StevenDoesStuffs** [comment]
I finished implementing what you've said above.

> Also, the default env https://github.com/nushell/nushell/blob/main/crates/nu-utils/src/sample_config/default_env.nu should reflect change NU_LIB/PLUGIN_DIRS to a constant.

I can't quite do this yet since currently there's no way of referencing `$nu.config-path` or doing path manipulation in consts.

File: crates/nu-parser/src/parse_keywords.rs (hunk 1 above)
Comment (jntrnr) [reply crates/nu-parser/src/parse_keywords.rs:4067]: yes, please fix that. I originally had it in there for debugging purposes and forgot to fix it.

**Comment by kubouch** [comment]
I see, then it's too soon to merge into two files. Let's keep the constant NU_LIB_DIRS and `$env.NU_LIB_DIRS` working side by side until we're ready to fully migrate.

**Comment by StevenDoesStuffs** [comment]
This is ready to merge now BTW.

File: tests/shell/mod.rs [hunk 2]
```diff
@@ -140,6 +140,36 @@ fn nu_lib_dirs_relative_repl() {
...
+                r#"
+                    const NU_LIB_DIRS = [ 'scripts' ]
+                    source-env foo.nu
+                    $env.FOO
+                "#,
+            )]);
+
+        let inp_lines = &[r#"source main.nu"#];
+
+        let actual_repl = nu!(cwd: dirs.test(), nu_repl_code(inp_lines));
```
Comment (kubouch) [inline tests/shell/mod.rs:166]: This could be tested using normal execution, REPL simulator should not be necessary.

Comment (StevenDoesStuffs) [reply tests/shell/mod.rs]: Fixed

**Comment by fdncred** [comment]
@StevenDoesStuffs can you resolve these conflicts again please?

**Comment by StevenDoesStuffs** [comment]
@fdncred done.

**Comment by fdncred** [comment]
Seems like we're ready. Thanks!

**Comment by fdncred** [comment]
@StevenDoesStuffs How does this need to change in env.nu to get it to work? We neglected to update the default configs.

**Comment by fdncred** [comment]
I'm thinking we should revert until we can get it working like it was in env.nu, but also allow const usage.

**Comment by fdncred** [comment]
This was reverted because we'd like to have our cake and eat it too. So this would still work,
```
let-env NU_LIB_DIRS = [
    ($nu.config-path | path dirname | path join 'scripts')
    'C:\Users\username\source\repos\forks\nu_scripts'
    ($nu.config-path | path dirname)
]
```
but also the const stuff created in this PR.

**Comment by StevenDoesStuffs** [comment]
This PR is already backwards compatible, so `let-env NU_LIB_DIRS` should already still work.

**Comment by fdncred** [comment]
> This PR is already backwards compatible, so `let-env NU_LIB_DIRS` should already still work.

Unfortunately, I didn't find that to be the case with my testing. Any scripts that I had in those 3 folders were "not found" when I tried running them via config.nu.

To be more verbose, nu_lib_dirs is defined in env.nu and this is a section of my config.nu

```
# Custom Commands
source defs.nu

# prompt
use prompt\oh-my.nu git_prompt
let-env PROMPT_COMMAND = {|| (git_prompt).left_prompt }
let-env PROMPT_COMMAND_RIGHT = {|| (git_prompt).right_prompt }

# external completions
use custom-completions\cargo\cargo-completions.nu *
use custom-completions\git\git-completions.nu *

# get_weather
use 'weather\get-weather.nu' get_weather
```

`defs.nu` was in the first item in nu_lib_dirs list.
the rest were in the second item in the nu_lib_dirs list.
neither of those worked.

Checked out your PR again just to test again and this is what I see when i startup nushell.
```
Error: nu::parser::module_not_found

  × Module not found.
     ╭─[C:\Users\username\AppData\Roaming\nushell\config.nu:992:1]
 992 │ # prompt
 993 │ use prompt\oh-my.nu git_prompt
     ·     ───────┬───────
     ·            ╰── module not found
 994 │ let-env PROMPT_COMMAND = {|| (git_prompt).left_prompt }
     ╰────
  help: module files and their paths must be available before your script is run as parsing occurs before anything is evaluated
```
That means that it doesn't see the second item from here
```
let-env NU_LIB_DIRS = [
    ($nu.config-path | path dirname | path join 'scripts')
    'C:\Users\username\source\repos\forks\nu_scripts'
    ($nu.config-path | path dirname)
]
```
When I try to run a custom command in my `defs.nu` file named `repeat`, I get
```
> repeat --help
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #2:1:1]
 1 │ repeat --help
   · ───┬──
   ·    ╰── did you mean 'reject'?
   ╰────
  help: program not found
```
The odd thing about it is that this looks right.
```
> $env.NU_LIB_DIRS
╭───┬─────────────────────────────────────────────────────╮
│ 0 │ C:\Users\username11\AppData\Roaming\nushell\scripts │
│ 1 │ C:\Users\username11\source\repos\forks\nu_scripts   │
│ 2 │ C:\Users\username11\AppData\Roaming\nushell         │
╰───┴─────────────────────────────────────────────────────╯
```
If I now try this in the repl, now that everything is started up, I get this
```
> use prompt\oh-my.nu git_prompt

  × Module not found.
   ╭─[entry #3:1:1]
 1 │ use prompt\oh-my.nu git_prompt
   ·     ───────┬───────
   ·            ╰── module not found
   ╰────
  help: module files and their paths must be available before your script is run as parsing occurs before anything is evaluated
```

**Comment by StevenDoesStuffs** [comment]
Huh that's not what's supposed to happen. I'll look into that soon. One thing is that if defined, const `NU_LIB_DIRS` will take priority over env `NU_LIB_DIRS`, so can you please check that that's not what's happening here?

**Comment by fdncred** [comment]
> Huh that's not what's supposed to happen. I'll look into that soon. One thing is that if defined, const `NU_LIB_DIRS` will take priority over env `NU_LIB_DIRS`, so can you please check that that's not what's happening here? I suppose that makes this PR technically not fully backwards compatible.

It wasn't what was happening. I only had one or the other defined at a time and the other one was commented out.

Thanks for looking into this.
